### PR TITLE
Suggestion(plugins): allow checking for other plugins while loading data

### PR DESCRIPTION
**Pull request by RisingLeaf**
**Feature:** This PR implements the feature request detailed and discussed in issue #9377

## Feature Details
Adds preparsers(same thing as c preprocessor) to ES data loading, with only one testable condition so far. They allow to only load a universe object(missions, ships, everything else) when a certain condition set is accepted.

I chose this syntax:
```
?IF
    condition a

<define something>

?ELSE

?IF
    condition b

<define something else>

?ENDIF

?ENDIF
```
(I basically copied from C, just less features)

The data will stop loading from the point where the line is placed, it works like any other top level token.
it can not be placed inside the definition of another top-level token like this:
```
mission something
    name bla
    job
?IF
    "plugin active: omnis"
    anothermissiontag
```
as that would break the definition of the mission

The preparser conditions use their own conditions store, as the player is not loaded at that point yet, as of now it only has one condition, if a plugin is installed.

## Testing Done
installed a random plugin, added the condition to a random es data file, depending on if I wrote the plugin name right the data was loaded, if I wrote another random name, no data was loaded

## Performance Impact
N/A

### CODE REVIEW

**Comment by DarcyManoel** [comment]
This is by no means meant as a lure to get this merged. BUT the use-case is blatant that plug-ins could easily support others without sacrificing their clean error files.

Like Omnis, I will personally support any plug-in within Omnis if it meant I could avoid throwing unnecessary errors through it all.

**Comment by RisingLeaf** [comment]
Changed it so that it is more in line with what @tibetiroka  thought of in #9377

**Comment by RisingLeaf** [comment]
I updated the description if you want to see what I changed

**Comment by RisingLeaf** [comment]
I will be so bold and mark this as ready for review as I am pretty happy with how it is rn

**Comment by quyykk** [comment]
No front but I find the syntax to be a bit ugly. I think making it something simple makes sense and we shouldn't be trying to make it as powerful as the C preprocessor.

The first suggestion mentioned in the issue is pretty neat. Thoughts? (not saying you must do it like this, but that's just my opinion).

**Comment by tibetiroka** [comment]
Yeah, that solution does look a lot simpler, but it also leaves less room for future expansion.

The current version uses the existing condition syntax. That already allows us to use this for pretty much anything, not just plugin compatibility fixes.

For instance, this implementation makes it possible to literally count how many times you have opened the game (and did something in at least one save file), which could be used by some plugins to create, for example, easter eggs for people using the plugin for a long time. This would not be possible otherwise, because even the playtime counter is a per-save mechanic.

Also, I'm pretty sure we could use this to optimize the game. If I recall correctly, we currently check every (spaceport) mission's conditions whenever you are visiting the spaceport, which could take a while. With this PR, we could disable entire story branches if you went the other way (or just failed a mission), and have the game do less work from the next time you start it - we wouldn't need to check the FW and Syndicate missions if you are with the Navy (once those storylines exist, that is).

Those are just some random things I could think of off the top of my head, but I'm sure it would have other uses as well. Maybe this could be used to better patch vanilla mission chains without needing separate events to fix stuff?

~~Also, this PR is 46 lines. How much simpler do you want it to be? I find using the existing condition syntax to be a simpler solution.~~

**Comment by tibetiroka** [comment]
Oh, I just realized you can't use global conditions in this syntax. Maybe that's something worth looking into? (It would enable doing cool stuff like the first idea.)

And yeah, after putting a bit more thought into it, the second idea wouldn't work in the current implementation either, because it would need the conditions to be re-evaluated for each save file loaded/pilot created. (That's something that could be worth profiling, though.)

**Comment by RisingLeaf** [comment]
Adding global conditions is easy, so I could definitely do it, but I wont until I get affirmation that this is a way that will be accepted.

File: source/UniverseObjects.cpp [hunk 1]
```diff
@@ -334,9 +353,36 @@ void UniverseObjects::LoadFile(const string &path, bool debugMode)
 	if(debugMode)
 		Logger::LogError("Parsing: " + path);
 
+	bool active = true;
```
Comment (Amazinite) [inline source/UniverseObjects.cpp:356]: We should probably print a warning if the end of the file is reached and active is false, i.e. we encountered an ?IF with no ?ELSE or ?ENDIF.

**Review by Amazinite — COMMENTED** [review]
I feel like this method of allowing for plugin content to optionally load is a bit too cumbersome, while at the same time not being fine-grain enough.
* What happens if we have multiple files that we want to optionally load? Are we adding these if/endif tokens to the top and bottom of every such file? While that would work, it feels somewhat cumbersome. Using these tokens to optionally load large parts of a plugin is also something that a plugin couldn't easily display to the player.
* What happens if we want to only optionally load a portion of a root node? Well we can't, because these preparsing tokens can't be used within a root node. This means that in order to have a portion of a root node's contents change depending on the installed plugins, you'd need to copy the entire root node and make the change in each optionally loaded section.

During our dev chat this week, @warp-core suggested making use of #9496 to allow for a simpler approach to optionally loading large sections of plugin content (large meaning on a per root node basis, as opposed to optionally loading within root nodes). We could have plugins check their dependencies list for if the dependent plugins are loaded, and if not, don't load the entire plugin. But say you have only a portion of your plugin that is dependent on another; you wouldn't want to disable the entire thing. In that case, we could allow for the creation of a "components" folder within the plugin folder, which could contain sub-plugins with their own plugin.txt files and their own dependencies. These sub-plugins would then contain files with the root nodes that one wishes to optionally load based on the presence of certain plugins. By being sub-plugins with plugin.txt files, this also provides us a way to display the components of a plugin and their dependencies in the plugins menu in game, something that using ?IF/?ENDIF on a whole file or series of files wouldn't allow us to do.

Due to the fact that these preparsing tokens can't optionally load data within a root node means that the above suggestion fully covers the use case of these tokens. Although, I would be open to the idea of allowing such tokens to be used within root nodes, therefore allowing for extremely fine-grain optional loading--I can imagine a mission gifting the player a different ship depending on if a certain plugin is installed, and preparsing tokens within root nodes means you could do this without having to copy the entire mission--but that would be quite the undertaking and would require us to refactor all of our Load functions, which likely warrants a lot more discussion as to how we'd implement such a feature.

Thoughts?

**Comment by RisingLeaf** [comment]
I can see that loading large parts of a plugin over a dependencies file would make more sense, yes.

> Due to the fact that these preparsing tokens can't optionally load data within a root node means that the above suggestion fully covers the use case of these tokens. Although, I would be open to the idea of allowing such tokens to be used within root nodes, therefore allowing for extremely fine-grain optional loading--I can imagine a mission gifting the player a different ship depending on if a certain plugin is installed, and preparsing tokens within root nodes means you could do this without having to copy the entire mission--but that would be quite the undertaking and would require us to refactor all of our Load functions, which likely warrants a lot more discussion as to how we'd implement such a feature.

I wouldn't do something like this in the Load Functions, but in the function that loads files into datanode structure as this would be a centralized place where all files go through, it would be a bit messy to implement, but possible.

In any case I will close this PR, as it doesn't do what you want and discussion should be held on the issue

**Pull request closed without merging**